src/tmpfiles.c

```diff
@@ -23,12 +23,99 @@
 
 #include "config.h"		/* Generated by configure script */
 
+#include <dirent.h>
+#include <ftw.h>
+#include <glob.h>
+#include <libgen.h>
+#include <grp.h>
+#include <pwd.h>
 #include <sys/sysmacros.h>
-#include "finit.h"
+
+#include <stdlib.h>
+#include <string.h>
+#ifdef _LIBITE_LITE
+# include <libite/lite.h>
+#else
+# include <lite/lite.h>
+#endif
+
 #include "helpers.h"
-#include "tmpfiles.h"
+#include "log.h"
 #include "util.h"
 
+int debug;
+
+int mksubsys(const char *dir, mode_t mode, char *user, char *group)
+{
+	mode_t omask;
+	int uid, gid;
+	int rc = 0;
+
+	omask = umask(0);
+
+	uid = getuser(user, NULL);
+	if (uid >= 0) {
+		gid = getgroup(group);
+		if (gid < 0)
+			gid = 0;
+
+		rc = makedir(dir, mode);
+		if (rc && errno == EEXIST)
+			rc = chmod(dir, mode);
+		if (chown(dir, uid, gid))
+			err(1, "Failed chown(%s, %d, %d)", dir, uid, gid);
+	}
+
+	umask(omask);
+
+	return rc;
+}
+
+/* basename(3) replacement that does not modify its argument */
+const char *basenm(const char *path)
+{
+	const char *basename;
+
+	basename = rindex(path, '/');
+	if (!basename)
+		return path;
+
+	basename++;
+	return basename;
+}
+
+int getuser(char *username, char **home)
+{
+#ifdef ENABLE_STATIC
+	if (home)
+		*home = "/";	/* XXX: Fixme */
+	return fgetint("/etc/passwd", "x:\n", username);
+#else
+	struct passwd *usr;
+
+	if (!username || (usr = getpwnam(username)) == NULL)
+		return -1;
+
+	if (home)
+		*home = usr->pw_dir;
+	return  usr->pw_uid;
+#endif
+}
+
+int getgroup(char *group)
+{
+#ifdef ENABLE_STATIC
+	return fgetint("/etc/group", "x:\n", group);
+#else
+	struct group *grp;
+
+	if ((grp = getgrnam(group)) == NULL)
+		return -1;
+
+	return grp->gr_gid;
+#endif
+}
+
```

Comment:
this is now duplicated code from the core of `finit` - should it be pulled into separate `.c/h` files for proper code reuse?

Reply:
Yeah it definitely should, no code duplication please :smiley: